### ISSUE TITLE
Can't close game window unless FlipScreen() is called

### ISSUE THREAD

**Issue by DaVince** [spheredev/neosphere#34]
This will probably be a very obscure issue, but I encountered it during a test.

In my small test application, I draw some text and FlipScreen() it only once. Then I make the game go into an endless loop that doesn't call on FlipScreen() at any point. What happens now is that I cannot close minisphere: I have to force kill it. The game hasn't crashed, it just doesn't ever send a "window close button clicked" signal to act appropriately.

Like I said, this will be obscure, but who knows there will be people who don't update what's displayed much at all (for example, a Sphere-based server of some sort).

**Comment by fatcerberus** [comment]
I would be surprised if this weren't the case in Sphere as well.  It's not an easy problem to solve as, if you don't FlipScreen every so often, control never leaves the JS engine.  Allegro runs a separate thread for the window procedure, which prevents a total lockup, but won't do anything in response to a close event--you have to handle that explicitly.  If `FlipScreen()` is never called, that handling can't happen.

In theory I could do what Allegro does and run an event pump in another thread, but I don't think you can longjmp across a thread boundary.  I certainly wouldn't want to try it--I can only see bad things coming from such shenanigans. :wink:

So the issue then becomes, what do we do with the close event?  Duktape doesn't have a function that lets you break execution immediately, and shutting down the JS engine during script execution invariably crashes it so I'd still be in the same boat.

**Comment by DaVince** [comment]
Okay, fair enough. I consider this such an edge case that we might as well just recommend proper use of FlipScreen() in any Sphere-related documentation. :P

**Comment by fatcerberus** [comment]
There's a slight workaround if you don't want to incur the performance hit of a backbuffer flip: Call `Delay(0)`.  `Delay` is guaranteed to do at least one cycle of the event pump, so that should allow you to close the window.

That said, `FlipScreen` with a set framerate is preferable, as it will minimize CPU use (a `Delay(0)` loop will peg the CPU).

**Comment by fatcerberus** [comment]
As for documentation, it turns out I did add a note about this in `minisphere-api.txt`:

```
FlipScreen();

  Flips the backbuffer. Note that this also performs event loop processing, so
  if it is not called regularly, the engine will stop responding.
```

**Comment by DaVince** [comment]
Oh, heh. How about that. I guess we can consider this closed, then?

**Comment by fatcerberus** [comment]
Yeah, nothing to see here, move along. :wink:
